src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -59,19 +59,170 @@ int32_t GlobalizationNative_CompareStringNative(const uint16_t* localeName, int3
         currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
     }
 
-    NSString *firstString = [NSString stringWithCharacters: lpStr1 length: cwStr1Length];
-    NSString *secondString = [NSString stringWithCharacters: lpStr2 length: cwStr2Length];
-    NSRange string1Range = NSMakeRange(0, cwStr1Length);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrComposed = sourceString.precomposedStringWithCanonicalMapping;
+    NSString *targetString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *targetStrComposed = targetString.precomposedStringWithCanonicalMapping;
+
+    NSRange comparisonRange = NSMakeRange(0, sourceStrComposed.length);
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return -2;
+
+    return [sourceStrComposed compare:targetStrComposed
+                              options:options
+                              range:comparisonRange
+                              locale:currentLocale];
+}
+
+NSString* ComposeString(NSString* str)
+{
+    NSString* source = str.precomposedStringWithCanonicalMapping;
+    // Below we are removing weightless characters from the string to get ICU behavior.
+    NSString* zarb = @"\u200d";
+    NSString* nullChar = @"\0";
+    // Remove zero width joiner
+    NSString* result = [source stringByReplacingOccurrencesOfString:zarb withString:@""];
+    // Remove null characters
+    result = [result stringByReplacingOccurrencesOfString:nullChar withString:@""];
+
+    return result;
+}
+
+/*
+Function:
+IndexOf
+*/
+Range GlobalizationNative_IndexOfNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpTarget, int32_t cwTargetLength,
+                                        const uint16_t* lpSource, int32_t cwSourceLength, int32_t comparisonOptions, int32_t fromBeginning)
+{
+    assert(cwTargetLength >= 0);
+    Range result = {-2, 0};
+
+    NSLocale *currentLocale;
+    if (localeName == NULL || lNameLength == 0)
+    {
+        currentLocale = [NSLocale systemLocale];
+    }
+    else
+    {
+        NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
+        currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
+    }
```

Comment:
We are doing this pretty often right? Could we possible have a helper function that can be used across all instances where we do this? 
`NSLocale *currentLocale = GetCurrentLocale(localeName, lNameLength);`

Reply:
Yes, that makes sense, will add.